File: c++/string/compressString.cc
#include <string>
#include <iostream>

using namespace std;

class Solution {
public:
    string compressString(string S) {
        string ret;
        if(S.size() == 0)
            return ret;
        size_t cnt = 0;
        char last_c;

        for(size_t i = 0; i < S.size(); ++i) {
            if(i == 0) {
                ++cnt;
                last_c = S[i];
            } else if(last_c == S[i]) {
                ++cnt;
            } else {
                ret.push_back(last_c);
                ret += std::to_string(cnt);
                cnt = 1;
                last_c = S[i];
            }
        }
        ret.push_back(last_c);
        ret += std::to_string(cnt);

        if(ret.size() < S.size())
            return ret;
        else
            return S;
    }
};

File: python/string/compressString.py
#!/usr/bin/env python3
#coding=utf-8

class Solution:
    def compressString(self, S: str) -> str:
        ret = str()
        if(len(S) == 0):
            return ret

        last_c = ''
        cnt = 0
        for idx, c in enumerate(S):
            if(idx == 0):
                last_c = c
                cnt += 1
            elif(c == last_c):
                cnt += 1
            else:
                ret += last_c
                ret += str(cnt)
                cnt = 1
                last_c = c
        return ret if len(ret) < len(S) else S

so = Solution()
ret = so.compressString("rrrrrLLLLLPPPPPPRRRRRgggNNNNNVVVVVVVVVVDDDDDDDDDDIIIIIIIIIIlllllllAAAAqqqqqqqbbbNNNNffffff")
print(ret)
